lightnode/kvstore/store.go

```diff
@@ -0,0 +1,41 @@
+package kvstore
+
+import (
+	"errors"
+	"github.com/syndtr/goleveldb/leveldb/iterator"
+)
+
+// Store implements a key-value store. May be backed by a database like LevelDB.
+type Store interface {
+
+	// Put stores the given key / value pair in the database, overwriting any existing value for that key.
+	Put(key []byte, value []byte) error
+
+	// Get retrieves the value for the given key. Returns a ErrNotFound error if the key does not exist.
+	Get(key []byte) ([]byte, error)
+
+	// Delete removes the key from the database. Does not return an error if the key does not exist.
+	Delete(key []byte) error
+
+	// DeleteBatch atomically removes a list of keys from the database.
+	DeleteBatch(keys [][]byte) error
+
+	// WriteBatch atomically writes a list of key / value pairs to the database. The key at index i in the keys slice
+	// corresponds to the value at index i in the values slice.
+	WriteBatch(keys, values [][]byte) error
+
+	// NewIterator returns an iterator that can be used to iterate over a subset of the keys in the database.
+	// Only keys with the given prefix will be iterated. The iterator must be closed by calling Release() when done.
+	// The iterator will return keys in lexicographically sorted order. The iterator walks over a consistent snapshot
+	// of the database, so it will not see any writes that occur after the iterator is created.
+	NewIterator(prefix []byte) (iterator.Iterator, error)
+
+	// Shutdown shuts down the store, flushing any remaining data to disk.
+	Shutdown() error
```

Comment:
Not sure why Shutdown and Destroy are needed explicitly, or if you have intended use cases

Reply:
Shutdown flushes the data out to disk and closes the database. Destroy goes a step further and deletes the on-disk data in addition to shutting things down. Destroy is mostly useful in test environments where you want to clean things up after a test.

Reply:
> Shutdown: I'd think this store is a durable one, so there is no extra step required to persist when it stops the store?

Even if there is no work required to make data durable, there are still resources that need to be released when a store is shut down. The following is taken from the godocs on level DB:

```
// Close closes the DB. This will also releases any outstanding snapshot,
// abort any in-flight compaction and discard open transaction.
```

> Destroy: can the test create a different store if the purpose is to have DB from scratch?

Yes, this is mostly just syntactic sugar for automatically cleaning up all resources held by the DB (including disk resources). If you don't like this method I can delete it, but it will add some repeated code to unit tests.